blockdevice/util/disk/disks.go

```diff
@@ -0,0 +1,283 @@
+	case "ssd":
+		return TypeSSD, nil
+	case "hdd":
+		return TypeHDD, nil
+	case "nvme":
+		return TypeNVMe, nil
+	case "sd":
+		return TypeSD, nil
+	}
+
+	return 0, fmt.Errorf("unknown disk type %v", id)
+}
+
+// Disk reresents disk information obtained by reading /sys/block.
+type Disk struct {
+	// Size disk size in bytes.
+	Size uint64
+	// Model from /sys/block/*/device/model.
+	Model string
+	// DeviceName device name (e.g. /dev/sda).
+	DeviceName string
+	// Name /sys/block/<dev>/device/name.
+	Name string
+	// Serial /sys/block/<dev>/device/serial.
+	Serial string
+	// Modalias /sys/block/<dev>/device/modalias.
+	Modalias string
+	// WWID /sys/block/<dev>/device/wwid.
+	WWID string
+	// UUID /sys/block/<dev>/device/uuid.
+	UUID string
+	// Type is the disk type: HDD, SSD, SD card, NVMe drive.
+	Type Type
+}
+
+// List returns list of disks by reading /sys/block.
+func List() ([]*Disk, error) {
+	disks := []*Disk{}
+
+	sysblock := "/sys/block"
+
+	devices, err := ioutil.ReadDir(sysblock)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read /sys/block directory %w", err)
+	}
+
+	for _, dev := range devices {
+		skip := false
+		deviceName := filepath.Base(dev.Name())
+
+		for _, prefix := range []string{"sg", "sr", "loop", "md", "dm-", "ram"} {
+			if strings.HasPrefix(deviceName, prefix) {
+				skip = true
+
+				break
+			}
+		}
+
+		if skip {
+			continue
+		}
+
+		disk := Get(deviceName)
+		if disk.Size == 0 {
+			continue
+		}
+
+		disks = append(disks, disk)
+	}
+
+	return disks, nil
+}
+
+// Get gathers disk information from sys block.
+func Get(dev string) *Disk {
+	sysblock := "/sys/block"
+
+	dev = filepath.Base(dev)
+
+	readFile := func(parts ...string) string {
+		path := filepath.Join(parts...)
+
+		f, e := os.Open(path)
+		if e != nil {
+			return ""
+		}
+
+		data, e := ioutil.ReadAll(f)
+		if e != nil {
+			return ""
+		}
+
+		return strings.TrimSpace(string(data))
+	}
+
+	blockSizeString := readFile(
+		fmt.Sprintf("/sys/class/block/%s/queue/logical_block_size", dev),
+	)
+	if blockSizeString == "" {
+		blockSizeString = "512"
+	}
+
+	var size uint64
+
+	s := readFile(sysblock, dev, "size")
+	if s != "" {
+		var err error
+
+		size, err = strconv.ParseUint(strings.TrimSpace(s), 10, 64)
+		if err != nil {
+			size = 0
+		}
+
+		blockSize, _ := strconv.ParseUint(strings.TrimSpace(blockSizeString), 10, 64) //nolint:errcheck
+
+		size *= blockSize
+	}
+
+	diskType := TypeUnknown
+	rotational := readFile(sysblock, dev, "queue/rotational")
+
+	switch {
+	case rotational == "1":
+		diskType = TypeHDD
+	case rotational == "0":
+		diskType = TypeSSD
+	case strings.Contains(dev, "nvme"):
+		diskType = TypeNVMe
+	case strings.Contains(dev, "mmc"):
+		diskType = TypeSD
+	}
+
+	return &Disk{
+		DeviceName: fmt.Sprintf("/dev/%s", dev),
+		Size:       size,
+		Model:      readFile(sysblock, dev, "device/model"),
+		Name:       readFile(sysblock, dev, "device/name"),
+		Serial:     readFile(sysblock, dev, "device/serial"),
```

Comment:
my understanding is that `serial,` `wwid` and `uuid` are not under `device/`

Reply:
sometimes they are, sometimes they're not. I'll read from both places.